crp_APL/controllers/ctrl_vehicle_control_lat_pure_p/src/ctrl_vehicle_control_lat_pure_p.cpp

```diff
@@ -0,0 +1,106 @@
+
+    m_traj_sub_ = this->create_subscription<autoware_planning_msgs::msg::Trajectory>("/plan/trajectory", 10, std::bind(&CtrlVehicleControlLat::trajCallback, this, std::placeholders::_1));
+    m_egoVehicle_sub_ = this->create_subscription<crp_msgs::msg::Ego>("/ego", 10, std::bind(&CtrlVehicleControlLat::egoVehicleCallback, this, std::placeholders::_1));
+
+    this->declare_parameter("/ctrl/lookahead_time", 1.0f);
+    this->declare_parameter("/ctrl/wheelbase", 2.7f);
+
+    RCLCPP_INFO(this->get_logger(), "ctrl_vehicle_control has been started");
+}
+
+void crp::apl::CtrlVehicleControlLat::trajCallback(const autoware_planning_msgs::msg::Trajectory input_msg)
+{
+    m_input.path_x.clear();
+    m_input.path_y.clear();
+    m_input.path_theta.clear();
+    double quaternion[4];
+
+    
+    // this callback maps the input trajectory to the internal interface
+    for (long unsigned int i=0; i<input_msg.points.size(); i++)
+    {
+        m_input.path_x.push_back(input_msg.points.at(i).pose.position.x);
+        m_input.path_y.push_back(input_msg.points.at(i).pose.position.y);
+    }
+
+    if (input_msg.points.size() > 0)
+        m_input.target_speed = input_msg.points.at(0).longitudinal_velocity_mps;
+    else
+        m_input.target_speed = 0.0f;
+
+}
+
+void crp::apl::CtrlVehicleControlLat::egoVehicleCallback(const crp_msgs::msg::Ego input_msg)
+{
+    m_input.vxEgo = input_msg.twist.twist.linear.x;
+    m_input.egoSteeringAngle = input_msg.tire_angle_front;
+
+    m_input.egoPoseGlobal[0] = input_msg.pose.pose.position.x;
+    m_input.egoPoseGlobal[1] = input_msg.pose.pose.position.y;
+}
+
+void crp::apl::CtrlVehicleControlLat::pure_p_control()
+{
+    
+    float distance_to_index = m_input.vxEgo * m_params.lookahead_time;
+
+    RCLCPP_INFO(this->get_logger(), "distance to index: %f", distance_to_index);
+
+    float current_distance = 0.0f;
+    int target_index = 0;
+
+    for (int i = 0; i < m_input.path_x.size()-1; i++)
+    {
+        current_distance += sqrt(pow(m_input.path_x.at(i+1) - m_input.path_x.at(i), 2) + pow(m_input.path_y.at(i+1) - m_input.path_y.at(i), 2));
+        if (current_distance > distance_to_index)
+        {
+            target_index = i;
+            break;
+        }
+    }
+
+    // calculate the steering angle
+    m_output.steeringAngleTarget = 0.0f;
+
+    float alpha = atan2(m_input.path_y[target_index], m_input.path_x[target_index]);
+
+    m_output.steeringAngleTarget = atan2(2.0f * m_params.wheelbase * sin(alpha) / distance_to_index, 1);
+
+}
+
+void crp::apl::CtrlVehicleControlLat::loop()
+{
+    // parameter assignments
+    m_params.lookahead_time = this->get_parameter("/ctrl/lookahead_time").as_double();
+    m_params.wheelbase = this->get_parameter("/ctrl/wheelbase").as_double();
+
+    // control algorithm
+    if(m_input.path_x.size() > 0 && m_input.path_y.size() > 0)
+        pure_p_control();
```

Comment:
else zero steering angle target? otherwise system keeps the last value which is not the safe behavior

Reply:
else i would just drop the msg (not publish anything), for ex. when the steering wheel is at 20 deg and all of a sudden comes back 0, is not the best behavior in my op...